trio/_core/_run.py

```diff
@@ -193,14 +179,10 @@ def __enter__(self):
         task = _core.current_task()
         if self._scope_task is not None:
             raise RuntimeError(
-                "cancel scope may not be entered while it is already "
-                "active{}".format(
-                    "" if self._scope_task is task else
-                    " in another task ({!r})".format(self._scope_task.name)
-                )
+                "cancel scopes may not be reused or reentered "
+                "(first used in task {!r})".format(self._scope_task.name)
             )
         self._scope_task = task
```

Comment:
Do we use `self._scope_task` for anything else now? If it's effectively a bool, it would probably be better to make an an actual bool, so we don't inadvertently hold onto any stack frames longer than we need to.

Reply:
We use it in `_exc_filter` (testing whether `self._scope_task._pending_cancel_scope() is self`). I also don't think holding onto a `Task` can actually leak much memory -- if the task is running, it's referenced by the runner so our reference doesn't change the memory usage, and if it's done, it doesn't have any stack frames so the memory usage is small. Is there something I'm missing?

Reply:
Maybe `CancelScope._exc_filter` should use `current_task()`? ...and maybe it should be refactored into `CancelScope._close`. And actually, now that all `Cancelled` exceptions are the same, we can hoist the check up out of `_exc_filter` entirely; no reason `_close` should call `MultiError.filter` at all if it's not going to catch anything. That's another matter though...

I guess you're right that coroutine objects drop their `cr_frame` reference when the coroutine exits. Huh, I didn't know that. That does reduce the likelihood of this reference causing problems. So now it still makes me nervous to hold onto an unnecessary reference to a complex object, but it's only on general principles, not because of any specific issue I can point to :-).